src/client.cpp

```diff
@@ -269,260 +269,99 @@ void ADDON_Destroy()
   m_CurStatus = ADDON_STATUS_UNKNOWN;
 }
 
-ADDON_STATUS ADDON_SetSetting(const char *settingName, const void *settingValue)
+ADDON_STATUS SetStringSetting(const string settingName, const char* settingValue, string currentValue, ADDON_STATUS returnValueIfChanged)
 {
-  string str = settingName;
-  if (str == "host")
+  string strSettingValue = settingValue;
+
+  if (strSettingValue != currentValue)
   {
-    string tmp_sHostname;
-    XBMC->Log(LOG_INFO, "%s - Changed Setting 'host' from %s to %s", __FUNCTION__, g_strHostname.c_str(), (const char*) settingValue);
-    tmp_sHostname = g_strHostname;
-    g_strHostname = (const char*) settingValue;
-    if (tmp_sHostname != g_strHostname)
-      return ADDON_STATUS_NEED_RESTART;
+    XBMC->Log(LOG_INFO, "%s - Changed Setting '%s' from %s to %s", __FUNCTION__, settingName.c_str(), currentValue.c_str(), strSettingValue.c_str());
+    currentValue = strSettingValue;
+    return returnValueIfChanged;
   }
-  else if (str == "webport")
+
+  return ADDON_STATUS_OK;
+}
+
+ADDON_STATUS SetIntSetting(const string settingName, int settingValue, int &currentValue, ADDON_STATUS returnValueIfChanged)
+{
+  if (settingValue != currentValue)
   {
-    int iNewValue = *(int*) settingValue;
-    if (g_iPortWeb != iNewValue)
-    {
-      XBMC->Log(LOG_INFO, "%s - Changed Setting 'webport' from %u to %u", __FUNCTION__, g_iPortWeb, iNewValue);
-      g_iPortWeb = iNewValue;
-      return ADDON_STATUS_NEED_RESTART;
-    }
+    XBMC->Log(LOG_INFO, "%s - Changed Setting '%s' from %d to %d", __FUNCTION__, settingName.c_str(), currentValue, settingValue);
+    currentValue = settingValue;
+    return returnValueIfChanged;
   }
-  else if (str == "use_secure")
+
+  return ADDON_STATUS_OK;
+}
+
+ADDON_STATUS SetBoolSetting(const string settingName, bool settingValue, bool &currentValue, ADDON_STATUS returnValueIfChanged)
```

Comment:
=> `ADDON_STATUS SetBoolSetting(const std::string& settingName, bool settingValue, bool &currentValue, ADDON_STATUS returnValueIfChanged)`

Reply:
Looking at the bodies of the three new functions, which are very similar, how about a template function?

```C++
namespace
{
template <typename T>
ADDON_STATUS SetSetting(const std::string& settingName, T settingValue, T& currentValue, ADDON_STATUS returnValueIfChanged)
{
  if (settingValue != currentValue)
  {
     XBMC->Log(LOG_INFO, "%s - Changed Setting '%s'", __FUNCTION__, settingName.c_str());
     currentValue = settingValue;
     return returnValueIfChanged;
  }
  return ADDON_STATUS_OK;
 }
} // unnamed namespace
```

Note that I have removed the logging of old and new value which would only be possible if you add a parameter for passing the right character for the `printf`of the value (which I consider ugly).

But again, this is just an idea idea. No need to follow my brain fart.

Reply:
Or use to_string to pass the values???

Then it’s completely the same. I’ll give it a try in the morning

Reply:
I had a look at this but doesn't work with extern C linkage. When I get around to a proper refactor I'll try it (Settings in own class etc).